### PR TITLE
adding getById support to resource collections that can support it

### PR DESCRIPTION
**Pull request by martinsawicki**
- adding an abstract ResourcesImpl base class to have all those collections inherit from
- ResourcesImpl requires the presence of getByGroup in the implementation
- Resourcesimpl provides getById automatically, implemented in terms of the individual getByGroup impl
- adding getById test to TestTemplate
- enforcing the presence of consistent createFluentModel across collections deriving from ResourcesImpl
- pushing converter logic and other shared functionlaity into base class
- fixing up GenericResources a bit

### CODE REVIEW

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/arm/collection/implementation/ResourcesImpl.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import com.microsoft.azure.management.resources.fluentcore.arm.collection.SupportsGettingById;
+
+/**
+ * Base class for resource collection classes.
+ * (Internal use only)
+ * @param <T> the individual resource type returned
+ * @param <ImplT> the individual resource implementation
+ */
+public abstract class ResourcesImpl<T, ImplT>
+    implements
```
Comment (anuchandy) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/arm/collection/implementation/ResourcesImpl.java]: ResourcesImpl or GroupableResourcesImpl ?

The items in this collection are groupable they extends from "GroupableResourceImpl"

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/arm/collection/implementation/ResourcesImpl.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    public abstract T getByGroup(String groupName, String name) throws CloudException, IOException;
+
+    @Override
+    public final T getById(String id) throws CloudException, IOException {
+        return this.getByGroup(
+                ResourceUtils.groupFromResourceId(id),
+                ResourceUtils.nameFromResourceId(id));
+    }
+
+    protected abstract ImplT createFluentModel(String name);
```
Comment (anuchandy) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/arm/collection/implementation/ResourcesImpl.java]: There is one more version of createFluentModel

```java
  createFluentModel(ImplT's inner type inner)
```

is it make sense to move that as well?

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/arm/collection/implementation/ResourcesImpl.java (hunk 1 above)
Comment (martinsawicki) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/arm/collection/implementation/ResourcesImpl.java]: great point, though  now in the process of addressing it I discovered that Deployments are weird - they're Groupable but they do not have a region, so they cannot inherit from GroupableResource right now. I'm looking into what should be one about that...

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/arm/collection/implementation/ResourcesImpl.java (hunk 2 above)
Comment (martinsawicki) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/arm/collection/implementation/ResourcesImpl.java]: yep, looking into that one as well..  I'll add some more checkins into this PR shortly.

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/arm/collection/implementation/ResourcesImpl.java (hunk 1 above)
Comment (anuchandy) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/arm/collection/implementation/ResourcesImpl.java]: Deployment is always seems a special case. just a thought, may be we can have 'DeploymentsImpl' not to inherit from 'ResourcesImpl' and have it implement getById() directly then we can impose bounding 
```java
  GrouapbleResourcesImpl<T extends GroupableResource, ImplT extends GroupableResourceImpl>

```

Comment (martinsawicki) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/arm/collection/implementation/ResourcesImpl.java]: yep, that's basically what I;m doing now. 
I think Resource/GroupableResource may need to be rethought a little, later

**Pull request merged**